src/typecheck.rs

```diff
@@ -244,6 +237,11 @@ pub fn typecheck_program(program: &Program) -> Result<(), TypeError> {
     let (input_var, action_block) = &program.action;
     let mut action_env = ctx.env.clone();
     if let Some(var) = input_var {
+        // Check if the input variable shadows any existing variable
+        if ctx.env.contains_key(var) {
+            error!("Input variable '{}' shadows an existing variable", var);
```

Comment:
don't log this, just returning the error is enough.

Reply:
It's already returned as an error, I don't want any logging for this either. See my comment to the other similar case.